docs/faq/env_var.md

```diff
@@ -56,6 +56,9 @@ export MXNET_GPU_WORKER_NTHREADS=3
     - NaiveEngine: A very simple engine that uses the master thread to do the computation synchronously. Setting this engine disables multi-threading. You can use this type for debugging in case of any error. Backtrace will give you the series of calls that lead to the error. Remember to set MXNET_ENGINE_TYPE back to empty after debugging.
     - ThreadedEngine: A threaded engine that uses a global thread pool to schedule jobs.
     - ThreadedEnginePerDevice: A threaded engine that allocates thread per GPU and executes jobs asynchronously.
+  - Note: ThreadedEngine and ThreadedEnginePerDevice are not thread-safe. Switch to using NaiveEngine
+          if you want to have multiple threads interacting with a single MXNet model at the same time.
```

Comment:
Suggestion for changes:
1. Change: threads => client threads, i.e.  if you want to have multiple client threads interacting with a single MXNet model
2. Add this: 
For example, if you use MXNet from a gunicorn HTTP server, set the MXNET_ENGINE_TYPE to NaiveEngine. The gunicorn HTTP server use a pre-fork worker model, i.e. it (pre) forks and keeps a group of processes running to handle various incoming HTTP requests and then forks additional processes if the load goes up. Since the fork of a process replicates the complete process address space including threads, keeping MXNet single-threaded via the use of NaiveEngine makes it safe.

Reply:
Good suggestion and example. But I'm not following the reasoning of the last sentence `Since the fork of a process replicates the complete process address space including threads, keeping MXNet single-threaded via the use of NaiveEngine makes it safe.`

The ThreadedEngine of MXNet schedules operations based upon the availability of NDArrays in their corresponding operation functions. The root cause for the problem we saw is that multiple threads pushes operations on NDArrays to the MXNet ThreadedEngine simultaneously. This will cause
1. data race, e.g. thread1 and thread2 writing to the same NDArray memory space.
2. dead lock, e.g. thread1 trying to copy from NDArray1 to NDArray2, while thread2 trying to copy from NDArray2 to NDArray1, and they are both waiting for the read on the source arrays ready.

If it's already multiple threads in a single process interacting with MXNet, does it matter by forking it or not?